source/common/router/upstream_request.h

```diff
@@ -184,15 +236,191 @@ class UpstreamRequest : public Logger::Loggable<Logger::Id::router>,
+    }
+    return {};
+  }
+  // If the filter manager determines a decoder filter has available, tell
+  // the router to resume the flow of data from downstream.
+  void onDecoderFilterBelowWriteBufferLowWatermark() override {
+    upstream_request_.onBelowWriteBufferLowWatermark();
+  }
+  // If the filter manager determines a decoder filter has too much data, tell
+  // the router to stop the flow of data from downstream.
+  void onDecoderFilterAboveWriteBufferHighWatermark() override {
+    upstream_request_.onAboveWriteBufferHighWatermark();
+  }
+
+  // These functions are delegated to the downstream HCM/FM
+  const Tracing::Config& tracingConfig() override;
+  const ScopeTrackedObject& scope() override;
+  Tracing::Span& activeSpan() override;
+  void resetStream() override;
+  Upstream::ClusterInfoConstSharedPtr clusterInfo() override;
+  Router::RouteConstSharedPtr route(const Router::RouteCallback&) override;
+  Http::Http1StreamEncoderOptionsOptRef http1StreamEncoderOptions() override;
+
+  // Intentional no-op functions.
+  void onResponseDataTooLarge() override {}
+  void onRequestDataTooLarge() override {}
+  void endStream() override {}
+  void disarmRequestTimeout() override {}
+  void resetIdleTimer() override {}
+  void onLocalReply(Http::Code) override {}
+  // Upgrade filter chains not supported.
+  const Router::RouteEntry::UpgradeMap* upgradeMap() override { return nullptr; }
+
+  // Unsupported functions.
+  void requestRouteConfigUpdate(Http::RouteConfigUpdatedCallbackSharedPtr) override {
+    IS_ENVOY_BUG("requestRouteConfigUpdate called from upstream filter");
+  }
+  void setRoute(Router::RouteConstSharedPtr) override {
+    IS_ENVOY_BUG("set route cache called from upstream filter");
+  }
+  void recreateStream(StreamInfo::FilterStateSharedPtr) override {
+    IS_ENVOY_BUG("recreateStream called from upstream filter");
+  }
+  void clearRouteCache() override { IS_ENVOY_BUG("clear route cache called from upstream filter"); }
+  void upgradeFilterChainCreated() override {
+    IS_ENVOY_BUG("upgradeFilterChainCreated called from upstream filter");
+  }
+
+  Http::RequestTrailerMapPtr trailers_;
+  Http::ResponseHeaderMapPtr informational_headers_;
+  Http::ResponseHeaderMapPtr response_headers_;
+  Http::ResponseTrailerMapPtr response_trailers_;
+  UpstreamRequest& upstream_request_;
+};
+
+// This is the last filter in the upstream filter chain.
+// It takes request headers/body/data from the filter manager and encodes them to the upstream
+// codec. It also registers the CodecBridge with the upstream stream, and takes response
+// headers/body/data from the upstream stream and sends them to the filter manager.
+//
+// Right now, the CodecFilter is implicitly configured as the last filter in the
+// chain and manually created. TODO(alyssawilk) fix this before removing alpha
+// on upstream filters.
+class CodecFilter : public Http::StreamDecoderFilter, public Logger::Loggable<Logger::Id::router> {
```

Comment:
Just as the above comment said, this filter encodes headers/data/data to the upstream.
` It takes request headers/body/data from the filter manager and encodes them to the upstream codec.`

But why this filter is the subclass of `Http::StreamDecoderFilter`, not the `Http::StreamEncoderFilter`?

We take headers/body/data from the filter manager, then we should call `encodeHeader`, `encodeData`, etc...

Reply:
I guess the reason is we defined `Http::StreamDecoderFilter` for the requesting process, `Http::StreamEncoderFIlter` for the response process. 

But It just feels strange put the word decoder/encoder in the upstream part.

Reply:
because decode filters handle request path.  encode filters _alter_ the response path. This is essentially the same  as the router filter which is a decide only filter.  but yeah agree the encode/decode codec vs encode/decode filter is confusing.